File: packages/core/src/rules/unique-ids.ts
import type { RequirementWithSource, ValidationError } from "../types.js";

/**
 * GRD-VALID-003: The tool shall validate that all requirement ids are unique
 * across the project. Duplicate ids are reported with the paths of the
 * conflicting files.
 */
export function uniqueIds(
  requirements: RequirementWithSource[]
): ValidationError[] {
  const errors: ValidationError[] = [];
  const byId = new Map<string, RequirementWithSource>();

  for (const r of requirements) {
    const existing = byId.get(r.id);
    if (existing) {
      errors.push({
        path: r.sourcePath,
        message: `Duplicate requirement id: ${r.id} (also in ${existing.sourcePath})`,
      });
    } else {
      byId.set(r.id, r);
    }
  }

  return errors;
}

File: crates/shallgraph-core/src/rules/unique_ids.rs
//! GRD-VALID-003: Requirement ids must be unique across the project.

use crate::types::{RequirementWithSource, ValidationError};

pub fn unique_ids(requirements: &[RequirementWithSource]) -> Vec<ValidationError> {
    let mut errors = Vec::new();
    let mut by_id: Vec<(&str, &RequirementWithSource)> = Vec::new();

    for r in requirements {
        if let Some(existing) = by_id.iter().find(|(id, _)| *id == r.id.as_str()) {
            errors.push(ValidationError::new(
                r.source_path.display().to_string(),
                format!(
                    "Duplicate requirement id: {} (also in {})",
                    r.id,
                    existing.1.source_path.display()
                ),
            ));
        } else {
            by_id.push((r.id.as_str(), r));
        }
    }

    errors
}
